text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1006 @@
+* just allocate memory,
+
+which are all not possible right now in const code, and some can't be done in Rust in const code at all.
+
+It should be possible to write `default` in a way that allows it to be called in const contexts
+for types whose `Default` impl's `default` method satisfies all rules that `const fn` must satisfy
+(including some annotation that guarantees this won't break by accident).
+It must always be possible to call `default` outside of const contexts with no limitations on the generic parameters that may be passed.
+
+Similarly it should be possible to write `compile_time_default` in a way that also requires calls
+outside of const contexts to only pass generic parameters whose `Default::default` method satisifies
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait impls
+
+It is now allowed to prefix a trait name in an impl block with `const`, marking that this `impl`'s type is now allowed to
+have methods of this `impl`'s trait to be called in const contexts (if all where bounds hold, like usual, but more on this later).
+
+An example looks as follows:
+
+```rust
+impl const Trait for Type {}
+```
+
+Such impls require that the trait is a `const trait`.
+
+All method bodies in a const trait impl are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+### Const traits
+
+Traits need to opt-in to being allowed to have const trait impls. Thus you need to declare your traits by prefixing the `trait` keyword with `const`:
+
+```rust
+const trait Trait {}
+```
+
+This in turn checks all methods' default bodies as if they were `const fn`, making them callable in const contexts.
+Impls can now rely on the default methods being const, too, and don't need to override them with a const body.
+
+We may add an attribute later to allow you to mark individual trait methods as not-const so that when creating a const trait, one can
+add (defaulted or not) methods that cannot be used in const contexts.
+
+It is possible to split up a trait into the const an non-const parts as discussed [here](#cant-have-const-methods-and-nonconst-methods-on-the-same-trait).
+
+All default method bodies of const trait declarations are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+Note that on nightly the syntax is
+
+```rust
+#[const_trait]
+trait Trait {}
+```
+
+and a result of this RFC would be that we would remove the attribute and add the `const trait` syntax.
+
+### Const trait bounds
+
+Any item that can have trait bounds can also have `const Trait` bounds.
```

Comment:
In rfcs#3729, we had a question about what the precedence should be for relaxed bounds - `?(const Sized)` or `(?const) Sized`? Are you opt-ing out of the constness or the whole trait?